### ISSUE TITLE
Allow returning results via the API

### ISSUE THREAD

**Issue by atkinsonm** [codeinspectorio/citool#1]
First, I just started checking out your tool today and I'm impressed so far. Great work!

I live by GitLab CI. All of my code runs through extensive automated tests and quality checks. What I would like to do is not just trigger a new build by running `code-inspector` but actually return the results via the same command or a follow on command. This would allow me to see the results without leaving GitLab and also to control the pass/fail behavior of my pipeline depending on the results returned. This is an essential feature for me because it would give me full ability to use `code-inspector` in my pipeline and put the results directly in the hands of developers so that they can be addressed.

**Comment by juli1** [comment]
We literally 💗 gitlab. And we would like to develop new features with through the API.

So, before we start piling features, few questions: once done, what do you exactly want to have? results include:
- diff from the previous run
- raw results (it can be very long and might end up with a giant json blob and lead to other issues)
- condensed results (number of violations, number of duplicates)
- the rating (A/B/C/...)

I believe that in the context you explained, you likely want to check that there is no regression with the previous run (e.g. you do not introduce more duplicate or violations). So you might want to read the previous analysis results as well. But I would like to hear your use-case because after all, this tool is developed for developers by developers :-)

Note that we are also developing a GraphQL endpoint that will let you query all the metrics using a simple GraphQL query (in case you are familiar with graphql) and a user-dedicated api key.

**Comment by atkinsonm** [comment]
Thanks for the quick reply!

Great questions. For my use case in particular, a simple output of the number of violations and duplicates might suffice. Assuming I could parse the results, even that high-level information could give me enough information to control the flow of the pipeline. For example I might fail on violations but allow to continue with warnings on duplicates. Alternatively I might just simply fail on any findings.

I could see the diff being useful for narrowing the scope to the changes in a particular branch only. I just want to be careful not to confuse 0 regressions with 0 _overall_ findings. Given the choice I'd probably use both: `--diff` mode with no failure tolerance on feature branches and all mode with warnings on master. That way I'd be able to control additional regressions while acknowledging that violations may already exist and may take time to address.

I will include the rating SVG image on my project as a badge so I don't anticipate ever needing to fetch the rating via the API.

The graphql capabilities sound interesting. I'm not familiar with graphql querying but I'm sure I could google my way through it :)

**Comment by juli1** [comment]
Thanks a lot for all these details. Give us few days and we will try to implement this features.
Likely, what we will produce is a JSON like this
```
{
 totalViolations: 10
 violationsPerLanguage: 
    {
      javascript: 2,
      python: 8
   }
 totalDuplicates: 2,
 totalDuplicatedLines: 100
}
```

That would let people some information about the language (for example, javascript is very sensitive). We can start with that and see later to implement on a diff.

Does that sound reasonable?

**Comment by atkinsonm** [comment]
Absolutely, thanks!

On Wed, Apr 24, 2019, 11:26 PM Julien <notifications@github.com<mailto:notifications@github.com>> wrote:

Thanks a lot for all these details. Give us few days and we will try to implement this features.
Likely, what we will produce is a JSON like this

{
 totalViolations: 10
 violationsPerLanguage:
    {
      javascript: 2,
      python: 8
   }
 totalDuplicates: 2,
 totalDuplicatedLines: 100
}


That would let people some information about the language (for example, javascript is very sensitive). We can start with that and see later to implement on a diff.

Does that sound reasonable?

**Comment by juli1** [comment]
Give us few days, I think we can provide this feature - I cannot guarantee it will be free of bugs at first :D

**Comment by juli1** [comment]
Hello,

We updated the code to provide a new functionality.
You will need to use it like this:
`code-inspector -w`

It will then show the number of violations and duplicates!

Thanks again for providing feedback, we hope you will enjoy our platform as much as we enjoy building it!

**Comment by juli1** [comment]
Also, you can see that the new python package is available. See https://pypi.org/project/code-inspector/#history

Please let us know if we can close this ticket.

**Comment by atkinsonm** [comment]
@juli1 that was quick! I have been trying it and noticing some interesting behavior:

1. When I use `code-inspector -w` alone, the `analysis_id` will increment from previous runs, but the violations and duplicates are always zero and the analysis does not appear in the analysis history on the website. Here is a result of clean output from a project with known violations: https://gitlab.com/jhctechnology/aws-cloudformation-utilities/awslint/-/jobs/203258552.
2. After a lot of troubleshooting, I discovered that `code-inspector --verbose -w` would _sometimes_ show the analysis results on the website, but I have not been able to retrieve successful JSON with violations or duplicates on known projects with both. Example: https://gitlab.com/jhctechnology/aws-cloudformation-utilities/awslint/-/jobs/203284041. That job does not show under analysis history, but I can construct the URL as https://www.code-inspector.com/analysis/result/71/11017. The analysis has zero results when the previous run had many: https://www.code-inspector.com/analysis/result/71/11014.

I created a Docker image and .gitlab-ci.yml template so I can easily include code-inspector in any of my GitLab CI projects with a single line. If you're interested:

- Docker image and GitLab CI template: https://gitlab.com/jhctechnology/code-quality/code-inspector
- Sample project which implements it: https://gitlab.com/jhctechnology/aws-cloudformation-utilities/awslint/-/jobs/203245571

Thanks again for the quick responses!

**Comment by atkinsonm** [comment]
Update: I found the source of my issue. code-inspector is only evaluating violations on my master branch, and the tests I was performing were in a feature branch. Once I merged the code into my master branch, the JSON started to show the results I was expecting: https://gitlab.com/jhctechnology/aws-cloudformation-utilities/awslint/-/jobs/203296755

I guess I have a second feature request for evaluating branches other than master. Given that the addition of the `-w` option in version 0.0.3 solves the original intent of this issue, I will be closing it. I will pick up the branches discussion in #2.
